zilliqa/src/consensus.rs

```diff
@@ -642,6 +658,11 @@ impl Consensus {
                     .insert_transaction_receipt(&block.hash(), &block_receipts)?;
             }
 
+            // Important - only add blocks we are going to execute because they can potentially
+            // overwrite the mapping of block height to block, which there should only be one of.
+            // for example, this HAS to be after the deal with fork call
+            self.add_block(block.clone())?;
```

Comment:
This is the original bug - the block was added to the db before the fork was dealt with. BUT, during the fork resolution, when it was getting head_block, this was the new block at that height, and not the proper head block.

The fix is to not add blocks to the db unless they are about to be executed. There is only one mapping of block height to block. Loose blocks are discarded.